### ISSUE TITLE
Has someone trained successfully?  The loss converges to 0.X  , but the segmentation effect is poor

### ISSUE THREAD

**Issue by zhxt0** [meetshah1995/pytorch-semseg#20]
model： fcn8s
default parameter,   n_epoch=50

when epoch=2，mean IoU=0.19
         epoch=5，meanIoU=0.425
         epoch=6,7,8....49,mean IoU acc and other metrics unchanged

Has someone trained successfully?

**Comment by meetshah1995** [comment]
I've trained it successfully to 61 mIoU using default settings, but without the polynomial learning rate schedule that was added recently.

Maybe try turning the scheduler off and then try? I'll try to share the trained models soon.

**Comment by zhxt0** [comment]
@meetshah1995 Thank you! 
I will try the default settings without the polynomial learning rate schedule, if it works well, I will close this issue.

**Comment by zhxt0** [comment]
The problem I encountered before  as shown below
**I used the saved model which epoch=10**
`python validate.py --model_path='fcn32s_pascal_1_10.pkl`
**The results:**
FreqW Acc : 	0.742302227386
Overall Acc: 	0.839702640426
Mean Acc : 	0.603257463541
Mean IoU : 	0.42762824508
0 0.837675757711
1 0.543137975043
2 0.169756356139
3 0.487439715047
4 0.437707177585
5 0.487330917381
6 0.632008734532
7 0.61375504545
8 0.620915214793
9 0.103603798388
10 0.0693262529399
11 0.272149482632
12 0.475228349911
13 0.333891558856
14 0.432487540528
15 0.561013944854
16 0.267951629557
17 0.454415833988
18 0.186504870779
19 0.576957398075
20 0.416935592499

**I used the saved model which epoch=24**
`python validate.py --model_path='fcn32s_pascal_1_24.pkl`
**results:**
FreqW Acc : 	0.742302227386
Overall Acc: 	0.839702640426
Mean Acc : 	0.603257463541
Mean IoU : 	0.42762824508
0 0.837675757711
1 0.543137975043
2 0.169756356139
3 0.487439715047
4 0.437707177585
5 0.487330917381
6 0.632008734532
7 0.61375504545
8 0.620915214793
9 0.103603798388
10 0.0693262529399
11 0.272149482632
12 0.475228349911
13 0.333891558856
14 0.432487540528
15 0.561013944854
16 0.267951629557
17 0.454415833988
18 0.186504870779
19 0.576957398075
20 0.416935592499

**The results are same**

The experiment without the polynomial learning rate schedule is still running...

**Comment by zhxt0** [comment]
@meetshah1995 You are right!
The reason is the polynomial learning rate schedule, when iter == max_iter, the learning rate becomes 0, so the training stop.

You can change the default max_iter(30000), or remove the polynomial learning rate schedule

**Comment by zhxt0** [comment]
@meetshah1995 
I implement alexnet-fcn based on AlexNet, I've trained it successfully to 46.7 mIoU using default settings without the polynomial learning rate schedule.
Would PR if you think it's appropriate, Thanks

**Comment by jiujing23333** [comment]
@meetshah1995 
the 61 mIoU is in training set?

**Comment by meetshah1995** [comment]
61 mIoU on validation set.

**Comment by jiujing23333** [comment]
@meetshah1995 
A great result. My score is poor, can you give me some suggestions? This is my parameters:
batch_size=16  ( 4 GPU)
data augment: random flipping.
lr = 1e-4, divided by 2 at every 10 epochs.
I have trained it for 40 epochs and the mIU is 51.9 now.

**Comment by meetshah1995** [comment]
I tried it with batch-size 1 for Pascal VOC, can you try that?

**Comment by monaj07** [comment]
@zhxt0 Can you make your FCN-alexnet implementation available?
Even though I switch off the lr_schedule, I still get a poor performance.
Thanks in advance!

**Comment by zhxt0** [comment]
@monaj07 I think you can comment `line 60 and 61` in `train.py`, and then train fcn-xs in the default setting, you will get ~60% mIoU on validation set.

And the FCN-Alexnet
```
import torch.nn as nn
import torch.nn.functional as F

class alexfcn(nn.Module):
    def __init__(self, n_classes=21, learned_billinear=False):
        super(alexfcn, self).__init__()
        self.learned_billinear = learned_billinear
        self.n_classes = n_classes

        self.features = nn.Sequential(
            nn.Conv2d(3, 64, kernel_size=11, stride=4, padding=100),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=3, stride=2),

            nn.Conv2d(64, 192, kernel_size=5, stride=1, padding=2),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=3, stride=2),

            nn.Conv2d(192, 384, kernel_size=3, stride=1),
            nn.ReLU(inplace=True),

            nn.Conv2d(384, 256, kernel_size=3, stride=1, padding=1),
            nn.ReLU(inplace=True),

            nn.Conv2d(256, 256, kernel_size=3, stride=1, padding=1),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=3, stride=2, ceil_mode=True),
        )
        self.classifier = nn.Sequential(
            nn.Dropout2d(),
            nn.Conv2d(256, 4096, kernel_size=6, stride=1),
            nn.ReLU(inplace=True),
            nn.Dropout2d(),
            nn.Conv2d(4096, 4096, kernel_size=1, stride=1),
            nn.ReLU(inplace=True),
            nn.Conv2d(4096, self.n_classes, kernel_size=1),
        )

        if self.learned_billinear:
            raise NotImplementedError

    def forward(self, x):
        conv = self.features(x)
        score = self.classifier(conv)

        out = F.upsample(score, x.size()[2:], mode='bilinear')

        return out

    def init_alex_params(self, alexnet, copy_fc8=True):

        blocks = [self.features]
        features = list(alexnet.features.children())

        for idx, conv_block in enumerate(blocks):
            for l1, l2 in zip(features, conv_block):
                if isinstance(l1, nn.Conv2d) and isinstance(l2, nn.Conv2d):
                    #print idx, l1, l2
                    assert l1.weight.size() == l2.weight.size()
                    assert l1.bias.size() == l2.bias.size()
                    l2.weight.data = l1.weight.data
                    l2.bias.data = l1.bias.data
        for i1, i2 in zip([1, 4], [1, 4]):
            l1 = alexnet.classifier[i1]
            l2 = self.classifier[i2]
            #print type(l1), dir(l1),
            l2.weight.data = l1.weight.data.view(l2.weight.size())
            l2.bias.data = l1.bias.data.view(l2.bias.size())
        n_class = self.classifier[6].weight.size()[0]
        if copy_fc8:
            l1 = alexnet.classifier[6]
            l2 = self.classifier[6]
            l2.weight.data = l1.weight.data[:n_class, :].view(l2.weight.size())
            l2.bias.data = l1.bias.data[:n_class]

```
You need to modify models/__init__.py to train FCN-alexnet

**Comment by monaj07** [comment]
@zhxt0 Thanks a bunch for the code! 
However, right now it is in `epoch=20` and it has been stuck in `mean-IoU=37~39%` for the past 10 epochs. I am not sure what prevents it from surpassing even 40%. 
Please note that `poly_lr_scheduler` is commented and I can get to 63% accuracy using `FCN32s`, but alexnet does not properly work for me.

**Comment by monaj07** [comment]
@meetshah1995 @ibadami Thanks for making these codes publicly available.
Have you tried training with a `batch_size` larger than one? Why does it degrade the accuracy? I wonder if the cross_entropy2D messes up the batch orderings.

**Comment by monaj07** [comment]
@zhxt0 Unfortunately it does not surpass 42% after 100 iterations, it is so confusing...
I literally did what you said, i.e. commenting those two lines, adding alexfcn class to the list of fcn models and modifying __init__ in model folder to call alexfcn and initialize it with Alexnet pre-trained weights properly.
Every thing else is in its default setting, including `lr=0.00001`. 
The weird thing is that I easily get to ~63% on FCN32s within 10 iterations!

**Comment by monaj07** [comment]
@zhxt0 I realized the problem is caused by Pascal dataloader. The SBD train data has some overlap with the validation set of Pascal VOC 2012 dataset. This leads to the extra performance. If you remove this overlap, you cannot achieve your previously obtained accuracy.
